src/main/java/org/jnode/fs/xfs/directory/BlockDirectoryEntry.java

```diff
@@ -46,63 +84,25 @@ public class BlockDirectoryEntry extends XfsObject {
     public BlockDirectoryEntry(byte[] data, long offset, boolean v5) {
         super(data, (int) offset);
         this.isV5 = v5;
-        isFreeTag = getUInt16(0) == 0xFFFF;
-        if (!isFreeTag()) {
-            nameSize = getUInt8(8);
-            iNodeNumber = getInt64(0);
-            byte[] buffer = new byte[nameSize];
-            System.arraycopy(data, (int) offset + 9, buffer, 0, nameSize);
+        freeTag = getUInt16(0) == 0xFFFF;
+
+        if (!freeTag) { // it is xfs_dir2_data_entry
+            iNodeNumber = readInt64();
+            nameLength = readUInt8();
+            byte[] buffer = new byte[nameLength];
```

Comment:
Just checking the usage of `FSUtils.toNormalizedString()` and it seems that most/all lines before that call that function allocate a buffer, then copy the string into that buffer then pass the buffer to that function. I think it should be possible to move that functionality into the `FSUtils.toNormalizedString()` function.
It would make `getOffsetSize()` cleaner and `toString()` would be more explicit about if the entry is being used or not.